firewood/src/db.rs

```diff
@@ -523,6 +525,117 @@ mod test {
+    //    /-> P1 - will get committed
+    // R1 --> P2 - will get dropped
+    //    \-> P3 - will get orphaned, but it's still known
+    async fn test_proposal_scope_historic() {
+        let db = testdb().await;
+        let batch1 = vec![BatchOp::Put {
+            key: b"k1",
+            value: b"v1",
+        }];
+        let proposal1 = db.propose(batch1).await.unwrap();
+        assert_eq!(&*proposal1.val(b"k1").await.unwrap().unwrap(), b"v1");
+
+        let batch2 = vec![BatchOp::Put {
+            key: b"k2",
+            value: b"v2",
+        }];
+        let proposal2 = db.propose(batch2).await.unwrap();
+        assert_eq!(&*proposal2.val(b"k2").await.unwrap().unwrap(), b"v2");
+
+        let batch3 = vec![BatchOp::Put {
+            key: b"k3",
+            value: b"v3",
+        }];
+        let proposal3 = db.propose(batch3).await.unwrap();
+        assert_eq!(&*proposal3.val(b"k3").await.unwrap().unwrap(), b"v3");
+
+        // the proposal is dropped here, but the underlying
+        // nodestore is still accessible because it's referenced by the revision manager
+        // The third proposal remains referenced
+        let p2hash = proposal2.root_hash().await.unwrap().unwrap();
+        assert!(db.all_hashes().await.unwrap().contains(&p2hash));
+        drop(proposal2);
+
+        // commit the first proposal
+        proposal1.commit().await.unwrap();
+        // Ensure we committed the first proposal's data
+        let committed = db.root_hash().await.unwrap().unwrap();
+        let historical = db.revision(committed).await.unwrap();
+        assert_eq!(&*historical.val(b"k1").await.unwrap().unwrap(), b"v1");
+
+        // the second proposal shouldn't be available to commit anymore
+        assert!(!db.all_hashes().await.unwrap().contains(&p2hash));
+
+        // the third proposal should still be contained within the all_hashes list
+        // would be deleted if another proposal was committed and proposal3 was dropped here
+        let hash3 = proposal3
+            .nodestore
+            .root_hash()
+            .unwrap()
+            .expect("should have a root hash");
+        assert!(db.manager.read().unwrap().all_hashes().contains(&hash3));
+    }
+
+    #[tokio::test]
+    // test that dropping a proposal removes it from the list of known proposals
+    // R1 - base revision
+    //  \-> P1 - will get committed
+    //   \-> P2 - will get dropped
+    //    \-> P3 - will get orphaned, but it's still known
+    async fn test_proposal_scope_orphan() {
+        let db = testdb().await;
+        let batch1 = vec![BatchOp::Put {
+            key: b"k1",
+            value: b"v1",
+        }];
+        let proposal1 = db.propose(batch1).await.unwrap();
+        assert_eq!(&*proposal1.val(b"k1").await.unwrap().unwrap(), b"v1");
+
+        let batch2 = vec![BatchOp::Put {
+            key: b"k2",
+            value: b"v2",
+        }];
+        let proposal2 = proposal1.clone().propose(batch2).await.unwrap();
```

Comment:
nit: not sure why the clone was needed here, but it's a test and it doesn't hurt anything.